### PR TITLE
modified useremail function to include banned users #6962

### LINKED ISSUE #6962: Include Emails of Banned Users in Moderator Email Search
Hi, this is a [first-timers-only issue](https://code.publiclab.org/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Moderators have access to a feature that allows them to lookup profiles by email.  During a recent attempt to unban a mistakenly banned user (#6883) it was discovered that this feature does not include the emails of banned users.

In discussion with the moderation team, there are some good use cases for including these emails:

1) if a user created their account a long time ago, they may not remember there username (or remember it correctly).  In those instances, finding a profile by email would be the only recourse.  

2) Including banned users may eliminate a round of back and forth if the person emailing the moderation team is using the email address they signed up with.

The current solution (related #7109)is to create a check box that when toggled will add banned from 2015 onward to the search.  We also need to add some text explaining what the check box does and the impact that using it may have on the site.

**NOTE:** _See the **update** section below for a link to the post in this conversation where the steps are located to complete the FTO._

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Update** 
The steps needed to complete the FTO can be found in the conversation below at [this post](https://github.com/publiclab/plots2/issues/6962#issuecomment-570727398).

[See this page](https://code.publiclab.org/#r=all) for some help in taking your first steps!

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### Please keep us updated

💬⏰ - We encourage contributors to be respectful to the community and provide an update **within a week** of claiming a first-timers-only issue. We're happy to keep it assigned to you as long as you need if you update us with a request for more time or help, but if we don't see any activity a week after you claim it we may reassign it to give someone else a chance. Thank you in advance! 

If this happens to you, don't sweat it! Grab another open issue.

### Is someone else already working on this?

🔗- We encourage contributors to link to the original issue in their pull request so all users can easily see if someone's already started on it. 

👥- **If someone seems stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://code.publiclab.org/#r=all). Thanks!

### 🤔❓ Questions?

Leave a comment below!

### PR DESCRIPTION
**Pull request by shipcy**
Fixes #6962   (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by nstjean** [comment]
@jywarren If you just want to double check that this is what you wanted!

**Comment by shipcy** [comment]
> This looks good! Just update the top of your post with the number of the original issue to create a link and I think this is all done.

I've done the changes. Is it okay now?

**Comment by nstjean** [comment]
Oh sorry! I meant here:

FireShot Capture 183 - modified useremail function to include banned users #6962 by shipcy ·_ - github com

**Comment by shipcy** [comment]
@nstjean all the changes are done!

**Comment by nstjean** [comment]
Looks like it's linking to itself! The original issue is 6962. :smile:

File: app/controllers/modified useremail function to include banned users [hunk 1]
```diff
@@ -65,15 +65,21 @@ def useremail
       if params[:address]
         # address was submitted. find the username(s) and return.
         @address = params[:address]
-        @users = User.where(email: params[:address])
-                 .where(status: [1, 4])
+        if params[:include_banned]
+          @users = User.where(email: params[:address]) 
+          .where('created_at > (?)', DateTime.new(2015)) # since 2015, whether banned or not
```
Comment (jywarren) [inline app/controllers/modified useremail function to include banned users]: Great!

**Comment by jywarren** [comment]
Hmm, i wonder why the tests failed. Let's look!

File: app/controllers/modified useremail function to include banned users [hunk 2]
```diff
@@ -65,15 +65,21 @@ def useremail
...
+        else
+          @users = User.where(email: params[:address]) 
+          .where(status: [1, 4]) 
+        end
+      else
+        # unauthorized. instead of return ugly 403, just send somewhere else
+        redirect_to '/dashboard'
       end
```
Comment (jywarren) [inline app/controllers/modified useremail function to include banned users]: Oh, take a look here. Is this `end` improperly nested? Shouldn't the `redirect_to` go after it?

Comment (shipcy) [reply app/controllers/modified useremail function to include banned users]: I've done the changes. Please have a look @jywarren

**Comment by nstjean** [comment]
It should look like this:
```
 def useremail 
   if logged_in_as(['admin', 'moderator']) 
     if params[:address] 
       # address was submitted. find the username(s) and return. 
       @address = params[:address] 
       if params[:include_banned]
         @users = User.where(email: params[:address]) 
           .where('created_at > (?)', DateTime.new(2015)) # since 2015, whether banned or not
       else
         @users = User.where(email: params[:address]) 
           .where(status: [1, 4]) 
       end
     end 
   else 
     # unauthorized. instead of return ugly 403, just send somewhere else 
     redirect_to '/dashboard' 
   end 
```

**Comment by nstjean** [comment]
Hmm, ok, I'm not sure! I will test out the code tomorrow to see what's going on!

File: app/controllers/modified useremail function to include banned users [hunk 3]
```diff
@@ -60,19 +60,24 @@ def reset_user_password
...
+       else
+         @users = User.where(email: params[:address]) 
+           .where(status: [1, 4]) 
+       end
+     end 
+   else 
+     # unauthorized. instead of return ugly 403, just send somewhere else 
+     redirect_to '/dashboard' 
+   end 
+
```
Comment (tsparksh) [inline app/controllers/modified useremail function to include banned users]:
```suggestion
  end
```

Comment (shipcy) [reply app/controllers/modified useremail function to include banned users]: Thank you. I'll change this asap.

**Comment by shipcy** [comment]
> Looks like you forgot the closing `end` for the function

@tsparksh I've done this. Thanks for helping me out.

**Comment by nstjean** [comment]
@tsparksh Whoops, my fault! Thank you for catching that!

**Comment by jywarren** [comment]
Ah, it looks like this needs a rebase now, over the latest master branch. Sorry, someone else must've edited the same file! 

You can do that by following the process outlined here, but you may be able to skip step 2 if your master is still in sync with publiclab's master: https://publiclab.org/wiki/contributing-to-public-lab-software#Rewinding+the+master+branch

In general there are more tips and guides on rebasing here: https://publiclab.org/wiki/developers#Resources

Thanks!

**Comment by shipcy** [comment]
@jywarren I didn't get you.

**Comment by nstjean** [comment]
@shipcy Did you make these changes by editing the file on GitHub, or did you make the changes on your computer and pushing them up with Git?

**Comment by skilfullycurled** [comment]
@nstjean, they were made using edit.  I'm pretty positive.

**Comment by nstjean** [comment]
Ok. @jywarren how do we rebase when they don't have the ability to use a terminal for this? Would one of us have to download the pull request to do it?

**Comment by skilfullycurled** [comment]
@scipcy, just pinging you over here.  Everyone is working on it now : ), it's really close!  

As an aside, a note on how to progress once you get to the pull request stage: the following things are _not things you could have known_ and we will add them to the tutorial so the next person will. 

When you're working alone:

1) Let your tests run.
2) If they fail: you can click on them to see why, it might be something easy.  If you don't know why they failed/what to do about it, @message (at)publiclab/connectors for help.  This is a group of people who (as the name implies) will either help you or find someone who can.  You can **_always_** @message this groups if you need help!
3) If they pass: @message (at)publiclab/reviewers.  This way they will know it's ready to be approved.

If you're working with a group of people, it's the same process except message the people from your group instead.  Often that will include someone who is more senior, and can help you fix why the tests aren't passing.  They may even have the authority to approve the pull request.

Start getting excited: you're almost there!! 

(I have to run to a meeting, probably will be gone for a few hours)

**Comment by jywarren** [comment]
> Ok. @jywarren how do we rebase when they don't have the ability to use a terminal for this? Would one of us have to download the pull request to do it?

Ah, good point. Yes, let's do that. Would one of you be able to? Otherwise I'm happy to do so! It looks like the GitHub web editor is not up to it at this point.

**Comment by nstjean** [comment]
@jywarren I certainly can! But if I do that will it have to get pushed up into a new PR? I don't think I can push straight up to this repo

**Comment by jywarren** [comment]
I think you may be able to... github has some helpful permissions for that. But if not, yes, we can link the new PR to this but @shipcy will still get credited!

**Comment by nstjean** [comment]
@jywarren I tried a rebase and found that admin_controller.rb is missing. It's not in the folder, it appears it got renamed to "modified useremail function to include banned users". Should I rename the file back and commit the change?

**Comment by jywarren** [comment]
Ah, yes, that would be great. Thank you!

**Comment by nstjean** [comment]
Hmm. I tried pushing the changes to `git push -f upstream shipcy-banned-users` and `git push origin shipcy-banned-users` and neither one changed this PR. But I can create a new PR from my branch.

**Comment by jywarren** [comment]
Hmm. uh oh. Needs another conflicts resolution. Gah! I'll do this on my laptop and upload. Sorry!!

**Comment by nstjean** [comment]
@jywarren I think it should be all set over in: https://github.com/publiclab/plots2/pull/7181

**Comment by jywarren** [comment]
Yeah! Got confused between this and that, but you're right and i merged it there. Thanks! And thanks to @shipcy for your patience! This was a tough process but I hope you feel good that it got resolved and your fix was merged (even if not in this PR)! 

Thanks for your help @shipcy and also @nstjean !!!

**Pull request closed without merging**